src/lib.rs

```diff
@@ -334,46 +393,62 @@ impl<'a> Extract<'a> {
+
+            match archive {
+                ArchiveKind::Plain(_) => {
+                    match fs::create_dir_all(into_dir) {
+                        Ok(_) => (),
+                        Err(e) => {
+                            if e.kind() != io::ErrorKind::AlreadyExists {
+                                return Err(Error::Io(e));
                             }
                         }
-                        let file_name = file_to_extract.file_name().ok_or_else(|| {
-                            Error::Update("Extractor source has no file-name".into())
-                        })?;
-                        let out_path = into_dir.join(file_name);
-                        let mut out_file = fs::File::create(&out_path)?;
-                        io::copy(&mut reader, &mut out_file)?;
-                    }
-                    ArchiveKind::Tar(_) => {
-                        let mut archive = tar::Archive::new(reader);
-                        let mut entry = archive
-                            .entries()?
-                            .filter_map(|e| e.ok())
-                            .find(|e| e.path().ok().filter(|p| p == file_to_extract).is_some())
-                            .ok_or_else(|| {
-                                Error::Update(format!(
-                                    "Could not find the required path in the archive: {:?}",
-                                    file_to_extract
-                                ))
-                            })?;
-                        entry.unpack_in(into_dir)?;
                     }
-                    _ => {
-                        panic!("Unreasonable code");
-                    }
-                };
+                    let file_name = file_to_extract
+                        .file_name()
+                        .ok_or_else(|| Error::Update("Extractor source has no file-name".into()))?;
+                    let out_path = into_dir.join(file_name);
+                    let mut out_file = fs::File::create(&out_path)?;
+                    io::copy(&mut reader, &mut out_file)?;
+                }
+                #[cfg(feature = "archive-tar")]
+                ArchiveKind::Tar(_) => {
+                    let mut archive = tar::Archive::new(reader);
+                    let mut entry = archive
+                        .entries()?
+                        .filter_map(|e| e.ok())
+                        .find(|e| e.path().ok().filter(|p| p == file_to_extract).is_some())
+                        .ok_or_else(|| {
+                            Error::Update(format!(
+                                "Could not find the required path in the archive: {:?}",
+                                file_to_extract
+                            ))
+                        })?;
+                    entry.unpack_in(into_dir)?;
+                }
+                #[allow(unreachable_patterns)]
+                _ => {
+                    panic!("Unreasonable code");
```

Comment:
this looks like a similar case to the `unreachable` above -- could you use either unreachable or panic in both places and add the same panic-message explaining how it's unreachable?